proposals/mmmm-async-let.md

```diff
@@ -0,0 +1,274 @@
+initialized by the expression on the right- hand side of the `=`. However, it
+differs in that the initializer expression is evaluated in a separate,
+concurrently-executing child task. The child task beginning running as soon as
+the `async let` is encountered. On normal completion, the child task will
+initialize the variables in the `async let`.
+
+Because the main body of the function executes concurrently with its child
+tasks, it is possible that the parent task (the body of `makeDinner` in this
+example) will reach the point where it needs the value of an `async let` (say,
+`veggies`) before that value has been produced. To account for that, reading a
+variable defined by an `async let` is treated as a potential suspension point,
+and therefore must be marked with `await`. When the expression on right-hand
+side of the `=` of an `async let` can throw an error, that thrown error will be
+propagated through the parent task when reading the variable, and therefore
+accessing the variable must be marked with some form of `try`.
+The task will suspend until the child task has completed initialization of the
+variable (or thrown an error), and then resume.
+
+Bringing it back to our example, note that the `chopVegetables()` function
+might throw an error if, say, there is an incident with the kitchen knife. That
+thrown error completes the child task for chopping the vegetables. The error
+will then be propagated out of the `makeDinner()` function, as expected. On
+exiting the body of the `makeDinner()` function with this error, any child
+tasks that have not yet completed (marinating the meat or preheating the oven,
+maybe both) will be automatically cancelled.  `async let` is similar to a
+`let`, in that it defines a local constant that is initialized by the
+expression on the right-hand side of the `=`. However, it differs in that the
+initializer expression is evaluated in a separate, concurrently-executing child
+task which runs as part of an implicit task group.  The child task can begin
+running as soon as the `async let` is encountered.  On normal completion, the
+child task will initialize the variables in the `async let`. The parent task
+can `await` the value of the variable, at which point it suspends execution
+until the result is available; if the child task can raise an error, then the
+parent task must `try` awaiting its value, and the child task's error will
+propagate through the parent.
+
+## Detailed design
+
+### `async let` as sugar to task groups
+
+`async let` can be desugared to task groups. Each `async let` declaration
+behaves as if it creates a new task group whose scope begins at the
+`async let` and extends to the end of the declaration's formal scope. The
+right-hand expression in the declaration is then `add`-ed as a child task.
+`await`-ing the value of one of the declared variables acts like invoking
+`next()` on the task group to await the result of its single child task, if
```

Comment:
this isn't quite right; `next()` returns values in _completion order_ not by specific identity.

We'll need `group.addWithHandle` basically.

Note also that we would NOT want to automatically _suddenly not submit async tasks if the parent task was cancelled_ as we have no way to fulfill an await on such variable if we did that, which IS the case with `group.add -> false (did not submit)`.